File: crypto/src/main/java/io/oversec/one/crypto/sym/SymmetricCryptoHandler.java
package io.oversec.one.crypto.sym;

import android.content.Context;
import android.content.Intent;
import io.oversec.one.crypto.AbstractEncryptionParams;
import io.oversec.one.crypto.BaseDecryptResult;
import io.oversec.one.crypto.EncryptionMethod;
import io.oversec.one.crypto.UserInteractionRequiredException;
import io.oversec.one.crypto.proto.Outer;
import io.oversec.one.crypto.sym.ui.SymmetricBinaryEncryptionInfoFragment;
import io.oversec.one.crypto.sym.ui.SymmetricTextEncryptionInfoFragment;
import io.oversec.one.crypto.symbase.BaseSymmetricCryptoHandler;
import io.oversec.one.crypto.symbase.SymmetricDecryptResult;
import io.oversec.one.crypto.ui.AbstractBinaryEncryptionInfoFragment;
import io.oversec.one.crypto.ui.AbstractTextEncryptionInfoFragment;

public class SymmetricCryptoHandler extends BaseSymmetricCryptoHandler {

    public static final int BCRYPT_FINGERPRINT_COST = 10;

    private final OversecKeystore2 mKeyStore;


    static {
        OversecKeystore2.noop();//init security provider
    }

    public SymmetricCryptoHandler(Context ctx) {
        super(ctx);
        mKeyStore = OversecKeystore2.getInstance(ctx);
    }

    @Override
    protected EncryptionMethod getMethod() {
        return EncryptionMethod.SYM;
    }


    @Override
    public AbstractEncryptionParams buildDefaultEncryptionParams(BaseDecryptResult tdr) {
        SymmetricDecryptResult r = (SymmetricDecryptResult) tdr;
        return new SymmetricEncryptionParams(r.getSymmetricKeyId(), null, null);
    }


    @Override
    public BaseDecryptResult decrypt(Outer.Msg msg, Intent actionIntent, String encryptedText) throws UserInteractionRequiredException {
        return tryDecrypt(msg.getMsgTextSymV0(), encryptedText);
    }

    @Override
    public AbstractTextEncryptionInfoFragment getTextEncryptionInfoFragment(String packagename) {
        return SymmetricTextEncryptionInfoFragment.newInstance(packagename);
    }


    @Override
    public AbstractBinaryEncryptionInfoFragment getBinaryEncryptionInfoFragment(String packagename) {
        return SymmetricBinaryEncryptionInfoFragment.newInstance(packagename);
    }


    @Override
    protected SymmetricKeyPlain getKeyByHashedKeyId(long keyhash, byte[] salt, int cost, String encryptedText) throws KeyNotCachedException {
        Long keyId = mKeyStore.getKeyIdByHashedKeyId(keyhash, salt, cost);
        return keyId == null ? null : mKeyCache.get(keyId);
    }

    @Override
    protected void handleNoKeyFoundForDecryption(long[] keyHashes, byte[][] salts, int costKeyhash, String encryptedText) throws UserInteractionRequiredException {
        //noop
    }

    @Override
    protected void setMessage(Outer.Msg.Builder builderMsg, Outer.MsgTextSymV0.Builder symMsgBuilder) {
        builderMsg.setMsgTextSymV0(symMsgBuilder);
    }


    public boolean hasAnyKey() {
        return !mKeyStore.isEmpty();
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/sym/SymmetricCryptoHandler.kt
package io.oversec.one.crypto.sym

import android.content.Context
import android.content.Intent
import io.oversec.one.crypto.AbstractEncryptionParams
import io.oversec.one.crypto.BaseDecryptResult
import io.oversec.one.crypto.EncryptionMethod
import io.oversec.one.crypto.UserInteractionRequiredException
import io.oversec.one.crypto.encoding.Base64XCoder
import io.oversec.one.crypto.proto.Outer
import io.oversec.one.crypto.sym.ui.SymmetricBinaryEncryptionInfoFragment
import io.oversec.one.crypto.sym.ui.SymmetricTextEncryptionInfoFragment
import io.oversec.one.crypto.symbase.BaseSymmetricCryptoHandler
import io.oversec.one.crypto.symbase.SymmetricDecryptResult
import io.oversec.one.crypto.ui.AbstractBinaryEncryptionInfoFragment
import io.oversec.one.crypto.ui.AbstractTextEncryptionInfoFragment

class SymmetricCryptoHandler(ctx: Context) : BaseSymmetricCryptoHandler(ctx) {

    private val mKeyStore: OversecKeystore2

    override val method: EncryptionMethod
        get() = EncryptionMethod.SYM

    init {
        mKeyStore = OversecKeystore2.getInstance(ctx)
    }


    override fun buildDefaultEncryptionParams(tdr: BaseDecryptResult): AbstractEncryptionParams {
        val r = tdr as SymmetricDecryptResult
        return SymmetricEncryptionParams(r.symmetricKeyId, Base64XCoder.ID, null)
    }


    @Throws(UserInteractionRequiredException::class)
    override fun decrypt(
        msg: Outer.Msg,
        actionIntent: Intent?,
        encryptedText: String?
    ): BaseDecryptResult {
        return tryDecrypt(msg.msgTextSymV0, encryptedText)
    }

    override fun getTextEncryptionInfoFragment(packagename: String): AbstractTextEncryptionInfoFragment {
        return SymmetricTextEncryptionInfoFragment.newInstance(packagename)
    }


    override fun getBinaryEncryptionInfoFragment(packagename: String): AbstractBinaryEncryptionInfoFragment {
        return SymmetricBinaryEncryptionInfoFragment.newInstance(packagename)
    }


    @Throws(KeyNotCachedException::class)
    override fun getKeyByHashedKeyId(
        keyhash: Long,
        salt: ByteArray,
        cost: Int,
        encryptedText: String?
    ): SymmetricKeyPlain? {
        val keyId = mKeyStore.getKeyIdByHashedKeyId(keyhash, salt, cost)
        return if (keyId == null) null else mKeyCache.get(keyId)
    }

    @Throws(UserInteractionRequiredException::class)
    override fun handleNoKeyFoundForDecryption(
        keyHashes: LongArray,
        salts: Array<ByteArray>,
        costKeyhash: Int,
        encryptedText: String?
    ) {
        //noop
    }

    override fun setMessage(
        builderMsg: Outer.Msg.Builder,
        symMsgBuilder: Outer.MsgTextSymV0.Builder
    ) {
        builderMsg.setMsgTextSymV0(symMsgBuilder)
    }


    fun hasAnyKey(): Boolean {
        return !mKeyStore.isEmpty
    }

    companion object {

        const val BCRYPT_FINGERPRINT_COST = 10

        init {
            OversecKeystore2.noop()//init security provider
        }
    }
}
